<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="ie=edge">
  <title>LocusFocus: A Colocalization Tool to Prioritize Genes and Tissues from GWAS and eQTL Integration</title>
  <link rel="shortcut icon" href="#" />
   <!-- Font Awesome -->
   <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.7.0/css/all.css">
   <!-- Bootstrap core CSS -->
   <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css" integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">
   <!-- Material Design Bootstrap -->
   <link href="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.7.4/css/mdb.min.css" rel="stylesheet">
   <!-- DataTables -->
   <link rel="stylesheet" href="{{ url_for('static', filename='css/jquery.dataTables.min.css') }}"/>
   <link rel="stylesheet" href="{{ url_for('static', filename='DataTables/datatables.min.css') }}"/>
   <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/style.css') }}"/>
   <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/d3Style.css') }}"/>
</head>

<body>
    <div class="container">
    <div class="row">
        <div class="col-md-6">
            <div class="card border-primary mb-3" style="max-width: 20rem;">
                <div class="card-header">Session ID</div>
                <div class="card-body">
                    <h4 class="card-title" id="sessionid"></h4>
                    <p class="card-text">Save the above string for your records to load or share your plot.</p>
                    <p class="card-text">Plots older than 7 days are deleted.</p>
                </div>
            </div>
        </div>
    </div>
    <hr/>
    <div class="row">
        <div class="col-md-12">
            <h3>Colocalization Plot</h3>
            <select id="selGene" onchange="optionChanged(this.value)"></select>
            <div id="plot_message"></div>
            <div id="plot"></div>
            <div class="alert alert-dismissible alert-info">
                <button type="button" class="close" data-dismiss="alert">&times;</button>
                <strong>Tissues selected but not drawn are due to unavailable eQTL data, likely due to little or no expression in that tissue</strong>
            </div>
        </div>
    </div>
    <hr/>
    <div class="row">
        <div class="col-md-12">
            <h3>Heatmap of Simple Sum Colocalization P-values</h3>
            <div id="heatmap"></div>
            <div class="alert alert-dismissible alert-info">
                <button type="button" class="close" data-dismiss="alert">&times;</button>
                <strong>Gray cells (with negative -log10 Simple Sum p-values) correspond to gene-tissue pairs with no Simple Sum (SS) P-value.<br>
                    The exact reason for the absence of a SS P-value can be gleaned in the table below </strong>
            </div>
        </div>
    </div>
    <hr/>
    <div class="row">
        <div class="col-md-12">
            <div id="table-area">
                <h3>Simple Sum -log10 P-values Table</h3>
                <table id="variants-table" class="table table-striped table-bordered table-condensed table-sm sortable" cellspacing="0" width="90%">
                    <thead>
                        <tr>
                            <th class="th-sm">Gene</th>
                        </tr>
                    </thead>
                    <tbody></tbody>
                </table>
            </div> <!-- table-area closing div -->
            <div class="alert alert-dismissible alert-info">
                <button type="button" class="close" data-dismiss="alert">&times;</button>
                <p><strong>-1 values correspond to gene-tissue pairs with no eQTL data (likely due to little or no expression)</strong></p>
                <p><strong>-2 values correspond to gene-tissue pairs with no significant eQTLs after Bonferroni correction</strong></p>
                <p><strong>-3 values correspond to gene-tissue pairs where the Simple Sum P-value computation failed, likely due to insufficient SNPs</strong></p>
            </div>
        </div> <!-- col-md-12 closing div -->
    </div> <!-- row closing div -->
    <hr/>
    </div> <!-- container closing div -->

    <!-- JS -->
  <script src="https://cdnjs.cloudflare.com/ajax/libs/d3/5.9.1/d3.js"></script>
  <!-- Plotly -->
  <script src="https://cdn.plot.ly/plotly-latest.min.js"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/plot.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/plot_heatmap.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='js/heatmap_table.js') }}"></script>
  <script type="text/javascript">
    var thefile="{{ sessionfile }}";
    var genesfile="{{ genesfile }}";
    var SSPvalues_file="{{ SSPvalues_file }}";
    var sessionid="{{ sessionid }}"
    function init() {
        d3.json("{{ url_for('static', filename=sessionfile) }}").then(response => {
            d3.json("{{ url_for('static', filename=genesfile) }}").then(genesResponse => {
                d3.json("{{ url_for('static', filename=SSPvalues_file) }}").then(SSResponse => {
                    d3.select("#sessionid").text(sessionid);
                    var data = response;
                    var genesdata = genesResponse;
                    populate_genes(genesdata, data)
                    plot_gwas(data, genesdata);
                    plot_heatmap(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
                    buildTable(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
                });
            });
        });
    }

    function populate_genes(genesdata, data) {
        var selector = d3.select("#selGene");
        var genenames = genesdata.map(gene => gene['name']);
        genenames.forEach((gene) => {
                selector
                    .append("option")
                    .text(gene)
                    .property('value', gene);
        });
    }

    function optionChanged(newgene) {
        var newurl=`/update/${sessionid}/${newgene}`
        d3.select('#plot').text("")
        d3.select('#plot_message').append("p").attr("class","text-warning").text(`Please wait while re-drawing eQTLs for gene ${newgene}`)
        // d3.json("{{ url_for('update_colocalizing_gene', session_id=sessionid, newgene=newgene) }}").then(response => {
        d3.json(newurl).then(newresponse => {
            d3.json("{{ url_for('static', filename=genesfile) }}").then(genesResponse => {
                var newdata = newresponse;
                var genesdata = genesResponse;
                plot_gwas(newdata, genesdata);
                d3.select('#plot_message').text("")
                // plot_heatmap(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
                // buildTable(SSResponse.Genes, SSResponse.Tissues, SSResponse.SSPvalues);
            });
        });
    }

    init();

  </script>
  <!-- JQuery -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <!-- Bootstrap tooltips -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.4/umd/popper.min.js"></script>
  <!-- Bootstrap core JavaScript -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.3.1/js/bootstrap.min.js"></script>
  <!-- MDB core JavaScript -->
  <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mdbootstrap/4.7.4/js/mdb.min.js"></script>
  <!-- DataTables -->
  <script type="text/javascript" src="{{ url_for('static', filename='js/jquery.dataTables.min.js') }}"></script>
  <script type="text/javascript" src="{{ url_for('static', filename='DataTables/datatables.min.js') }}"></script>

</body>


</html>
